NDTensors/src/dense/fill.jl

```diff
@@ -22,18 +22,27 @@ function generic_randn(StoreT::Type{<:Dense}, dim::Integer=0)
 end
 
 function generic_zeros(
-  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
+  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer
 ) where {DataT<:AbstractArray,ElT}
   @assert ElT == eltype(DataT)
   data = generic_zeros(DataT, dim)
   StoreT = set_datatype(StoreT, typeof(data))
   return StoreT(data)
 end
 
-function generic_zeros(StoreT::Type{<:Dense{ElT}}, dim::Integer=0) where {ElT}
+function generic_zeros(
+  StoreT::Type{<:Dense{ElT,DataT}}, dims
+) where {DataT<:AbstractArray,ElT}
+  @assert ElT == eltype(DataT)
+  data = generic_zeros(DataT, dim(dims))
+  StoreT = set_datatype(StoreT, typeof(data))
+  return StoreT(data)
+end
```

Comment:
These two definitions should be combined.

I also think it probably doesn't make sense to allow `dims` to be anything besides either `Int` or `Tuple{Int}`, since `Dense` is an `AbstractVector`.

So this could be something like:
```julia
function generic_zeros(
  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer
) where {DataT<:AbstractArray,ElT}
  return generic_zeros(StoreT, (dim,))
end

function generic_zeros(
  StoreT::Type{<:Dense{ElT,DataT}}, dims::Tuple{Integer}
) where {DataT<:AbstractArray,ElT}
  @assert ElT == eltype(DataT)
  data = generic_zeros(DataT, dim(dims))
  StoreT = set_datatype(StoreT, typeof(data))
  return StoreT(data)
end

function generic_zeros(
  StoreT::Type{<:Dense{ElT,DataT}}, dim::Tuple{Integer,Integer,Vararg{Integer}}
) where {DataT<:AbstractArray,ElT}
  return error("Can't make a multidimensional `Dense` object.")
end
```
though I'm not sure if we are expecting `dim`/`dims` to be something besides integers.

Is there code where `generic_zeros` is being called to construct a `Dense` where `dims` has length greater than 1? If so I would be curious where that is happening and why (maybe it is needed for generic code where that makes sense for another storage type?).

Reply:
Yes in `blocksparsetensor.jl` that calls generic_zeros with a tuple of length greater than one
```
if !Rblock_exists
      # Rblock doesn't exist
      block_size = permute(size(Tblock), perm)
      # TODO: Make GPU friendly.
      DenseT = set_type_parameters(Dense, (eltype, parenttype), (eltype(R), datatype(R)))
      Rblock = tensor(generic_zeros(DenseT, block_size), block_size)
elseif !Tblock_exists
      # Tblock doesn't exist
      block_size = permute(size(Rblock), invperm(perm))
      # TODO: Make GPU friendly.
      DenseT = set_type_parameters(Dense, (eltype, parenttype), (eltype(T), datatype(T)))
      Tblock = tensor(generic_zeros(DenseT, block_size), block_size)
 end
 ```

Reply:
I see, probably it is better to change that code rather than make `generic_zeros` more general, for example:
```julia
      Rblock = tensor(generic_zeros(DenseT, prod(block_size)), block_size)
```
etc.

Reply:
Okay I have flattened these and it looks like it might have been the only case. I also updated your suggestion to remove the `where{}` portion of the code. Let me know if it looks reasonable, thanks!

Reply:
Also if the code looks good I can update the `generic_randn` code too from this file